corehq/motech/openmrs/finders.py

```diff
@@ -0,0 +1,287 @@
+            if id_type_uuid in case_config_ids:
+                case_property = case_config_ids[id_type_uuid]['case_property']
+                value = identifier['identifier']
+                case_update[case_property] = value
+        case_block = CaseBlock(
+            case_id=case.get_id,
+            create=False,
+            update=case_update,
+        )
+        submit_case_blocks([case_block.as_string()], case.domain)
+
+    def find_patients(self, requests, case, case_config):
+        """
+        Matches cases to patients. Returns a list of patients, each
+        with a confidence score >= self.threshold
+        """
+        from corehq.motech.openmrs.logger import logger
+        from corehq.motech.openmrs.repeater_helpers import search_patients
+
+        self.set_property_map(case_config)
+
+        candidates = {}  # key on OpenMRS UUID to filter duplicates
+        for prop in self.searchable_properties:
+            value = case.get_case_property(prop)
+            if value:
+                response_json = search_patients(requests, value)
+                for patient in response_json['results']:
+                    score = self.get_score(patient, case)
+                    if score >= self.threshold:
+                        candidates[patient['uuid']] = PatientScore(patient, score)
+        if not candidates:
+            logger.info(
+                'Unable to match case "%s" (%s): No candidate patients found.',
+                case.name, case.get_id,
+            )
+            return []
+        patients_scores = sorted(candidates.values(), key=lambda cand: cand.score, reverse=True)
+        if len(patients_scores) == 1:
+            patient = patients_scores[0].patient
+            self.save_match_id(case, case_config, patient)
+            logger.info(
+                'Matched case "%s" (%s) to ONLY patient candidate: \n%s',
+                case.name, case.get_id, pformat(patient, indent=2),
+            )
+            return [patient]
+        if patients_scores[0].score / patients_scores[1].score > 1 + self.confidence_margin:
+            # There is more than a `confidence_margin` difference
+            # (defaults to 10%) in score between the best-ranked
+            # patient and the second-best-ranked patient. Let's go with
+            # Patient One.
+            patient = patients_scores[0].patient
+            self.save_match_id(case, case_config, patient)
+            logger.info(
+                'Matched case "%s" (%s) to BEST patient candidate: \n%s',
+                case.name, case.get_id, pformat(patients_scores, indent=2),
+            )
+            return [patient]
+        # We can't be sure. Just send them all.
+        logger.info(
+            'Unable to match case "%s" (%s) to patient candidates: \n%s',
+            case.name, case.get_id, pformat(patients_scores, indent=2),
+        )
+        return [ps.patient for ps in patients_scores]
```

Comment:
Looks like we are returning `None` if there are more than 1 matches, why are we returning everything then?

Reply:
The `find_patient()` function used by the OpenMRS Repeater gives a definitive answer -- either one OpenMRS patient, or `None` if it isn't sure. 

But PatientFinders can be more explicit, and that helps us to know what's going on. So if a PatientFinder doesn't find any patients, it returns an empty list. If it can confidently pick a patient (in the case of this `WeightedPropertyPatientFinder`, because only one passed the threshold, or only one had a score that was much higher than the rest) it must return just that one. But if it found many but can't pick one, we would want to know, and it should return all of them. 

I'm happy to separate the expected behaviour of `PatientFinder.find_patients()` from `repeater_helpers.find_patient()` because a PatientFinder is quite a different idea from the Repeater that uses it.

We might want to write better/different PatientFinders -- other projects might have different requirements. (I'm thinking a [LevenshteinPatientFinder](https://en.wikipedia.org/wiki/Levenshtein_distance) might be useful in some situations.) 

So we want to know if our PatientFinder is consistently letting us down. Finding no patients is fine, because the patient might not exist in OpenMRS, but finding a lot is a bad sign. (We might want to add a soft_assert in `find_patient()` to mail me when we get multiple results.) (... And it would be nice to have a unit test that mocks with copies of real data. Or compare multiple PatientFinder implementations against the same sample set of patients. :thinking: )

Reply:
@kaapstorm That makes sense if this would be used for different patient finders. But if thats not happening soon can consider the overhead of returning all of patients, unless these would always be a small number.

Reply:
They really *should* be a small number. Any project that uses PatientFinders will have to watch to see what is being returned, so they can tweak their config. If they are getting a lot of matches, then their weights are wrong.